AP_Scripting/Rover: support Lua motor drivers

rmackay9 opened the pull request:
This PR includes part of the walking robots PR https://github.com/ArduPilot/ardupilot/pull/14884 but is smaller allowing it to be more easily merged.

Similar to @IamPete1's PR https://github.com/ArduPilot/ardupilot/pull/14552, the controversial part of this change is exposing set_angle and set_range to scripting.  The request to add set_output_norm() is not straightforward though and becomes a significant blocker for the walking robots GSoC project.

This PR allows Lua scripts to implement motor drivers for Rovers.  The changes include:

- AP_Vehicle gets a new "get_control_output" function that allows the caller to retrieve the high level outputs of the controllers (attitude, steering, etc) in the range of -1 to +1
- Rover implements "get_control_output" for yaw (aka Steering) and throttle. In the case of Rover these control outputs are held within the Motors library as inputs.  I.e. retrieving the input from the motors library is the same as retrieving the outputs from the controllers.
- AP_Script gets bindings for AP_Vehicle's get_control_output
- AP_Script gets bindings for SRV_Channels's set_angle and set_range
- A new example script is added to AP_Scripting which demonstrates that this all works for a simple ackerman steering vehicle.  Below is a screen shot of the rover driving around successfully

![image](https://user-images.githubusercontent.com/1498098/89699747-ea0e0b00-d963-11ea-8c56-b8a30c0fd53d.png)

This has been tested in SITL by comparing the regular motor driver's outputs to throttle and steering channels to the scripts outputs for script1 and script3 and the outputs were identical.

IamPete1 commented:
I like this, such a scheme also makes scene for Plane I think. Maybe even copter... 

Rather than adding a special function you could also call get_output_scaled for throttle steering ect servo functions.

rmackay9 commented:
@IamPete1, yes, I think this should work for the other vehicles too.

The GSoC walking robots project originally used a method which passed the controller outputs via the servo outputs but the scripts then need to output PWM values to the servos using some of the same outputs which became messy.

IamPete1 commented:
We have servo outputs direct from the controller on plane (I think, maybe its copter)
https://github.com/ArduPilot/ardupilot/blob/e74acba4ecf8eed6cc294b29d62f915a4eeb0e94/libraries/SRV_Channel/SRV_Channel.h#L150

The Idea there was you could use a external gyro system, I would vote for adding more of that type of output and then lua becomes the 'external' gyro system. There is no need to actually assign them to a output for that to work.

rmackay9 commented:
@IamPete1, I can understand using the pwm interface for external systems that have no alternative but I think for Lua, where we can have much tighter integration there's no need to rely on the PWM interface.  We can do so much better..

WickedShell commented:
Based on our other discussions, I'm not seeing the difficulty with getting the scaled output set correctly to operate in the 0-1 range (at least not one that has any more risks then we already have), is there still a blocker?

tridge commented on Rover/Rover.cpp:
```diff
@@ -180,6 +180,32 @@ bool Rover::set_steering_and_throttle(float steering, float throttle)
+{
+    switch (control_output) {
+    case AP_Vehicle::ControlOutput::Throttle:
+        control_value = g2.motors.get_throttle() / 100.0f;
```
be nice to use constants for the 100 and 4500 values

rmackay9 commented:
As discussed I'll have another go at creating the set_output_normalised() using the range from the first output with the function assigned.

rmackay9 commented:
As discussed on last week's dev call I've reworked this slightly so that a new SRV_Channels::set_output_norm() is used.  I've retested and this works fine.  A couple of unhappy findings:

- set_output_norm is equivalent to calling set_output_pwm() meaning it bypasses the e-stop feature in calc_pwm
- the input to the function is always in the range -1 to +1 regardless of whether the channel is an "angle" or a "range".  This will be quite unintuitive for copter programmers who would expect a 0 to +1 range for the motor outputs.

![image](https://user-images.githubusercontent.com/1498098/90354533-921e8500-e084-11ea-9626-a717101654ec.png)

IamPete1 commented on libraries/SRV_Channel/SRV_Channel.cpp:
```diff
@@ -137,6 +137,18 @@ void SRV_Channel::set_output_pwm(uint16_t pwm, bool force)
+    // convert normalised value to pwm
+    const uint16_t mid = (servo_max + servo_min) / 2;
+    const uint16_t range = (servo_max - servo_min);
+    set_output_pwm(constrain_int16(mid + (value * range * 0.5f), servo_min, servo_max));
```
This whole function could be this?

```suggestion
    if (type_angle()) {
        pwm_from_angle(value * high_out)
    } else {
        pwm_from_range(value * high_out)
   }
    
```

rmackay9 replied:
@IamPete1,

Yes, thanks.  I think that would work too the only issue is that it would be inconsistent with the get_output_norm().  I'm reluctant to change get_output_norm() because that would affect the KDECAN ESCs and I'm keen that we get something in so that we can get the walking robots support merged before the end of the GSoC period.

IamPete1 replied:
I think your better of having a new function and leave the old one for KDECAN, for example, if I have a rover with a throttle 1000 is backwards, 1250 is stop and 2000 is forwards. So do set output norm from scripting with 0 and find I have 1/3rd throttle, and I have no way of finding what value I should use to stop.

rmackay9 replied:
@IamPete1, OK, thanks.  Maybe I'll implement it as you've suggested and then just add some comments to say that get_output_norm() and set_output_norm() are inconsistent.

rmackay9 replied:
@IamPete1, ah, the problem with this potential solution is that the high_out hasn't been set for the script1 outputs so we would need to add a set-range binding and then ask devs to call this and set it to something.  What they set it to wouldn't be too important but it would have to be set.

IamPete1 replied:
We could default to angle 4500, https://github.com/ArduPilot/ardupilot/blob/8484e8271f686e259559fba62a874275bc4307d0/libraries/SRV_Channel/SRV_Channel_aux.cpp#L129

But also give the bindings to change it

rmackay9 replied:
@IamPete1, I have never seen this defaulting before!  I had no idea it existed.

I'm happy with this idea, let's see what @WickedShell says.

WickedShell replied:
I'm fine with setting the default range on those to something, Personally I'd set it to range not angle, but I'm not that invested in which it goes to. I would suggest we can jump straight to using a much larger value then `4500` though and just give the scripting outputs more resolution off the bat :) (I realize our actual output layer is gonna truncate these down badly, but hey, one thing at a time :D )

rmackay9 replied:
OK, i've gone ahead and done this now and re-tested and it all seems OK. txs!

rmackay9 commented:
I've updated these changes with the following fixes:

1. fixed the example script so that it can handle reversing
2. added a constrain in Rover's get_control_output to ensure the values are always between -1 and +1 (sometimes the controllers can actually come up with a value above or below this)
3. defaulted the ranges for the k_scripting1 .. k_scripting16 servo outputs to be -4500 to +4500
4. added bindings for SRV_Channels::set_range and set_angle

I hope this is OK but all feedback welcome.  I'm quite keen to get this in because we only have a little over a week to finish off walking robots which is backed up behind this.

WickedShell commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -179,6 +179,7 @@ singleton AP_Vehicle method get_likely_flying boolean
 singleton AP_Vehicle method get_time_flying_ms uint32_t
 singleton AP_Vehicle method start_takeoff boolean float (-LOCATION_ALT_MAX_M*100+1) (LOCATION_ALT_MAX_M*100-1)
 singleton AP_Vehicle method set_target_location boolean Location
+singleton AP_Vehicle method get_control_output boolean AP_Vehicle::ControlOutput'enum AP_Vehicle::ControlOutput::Roll AP_Vehicle::ControlOutput::Last_ControlOutput float'Null
```
```suggestion
singleton AP_Vehicle method get_control_output boolean AP_Vehicle::ControlOutput'enum AP_Vehicle::ControlOutput::Roll AP_Vehicle::ControlOutput::Last_ControlOutput-1 float'Null
```

You don't want to allow access to Last_ControlOutput as it isn't actually a valid one to read.

Hwurzburg commented:
any chance of getting rough docs that I could polish and  add to upcoming features section?
